File: src/tools/remediateDrift.ts
import type { Tool, ToolProgress } from "src/Tool";
import { detectDrift } from "src/state/driftDetector";
import type { StackDiff } from "src/types/stack";
import { stringify as stringifyYaml } from "yaml";
import { z } from "zod";

export const RemediateDriftInputSchema = z.object({
  stackName: z.string().min(1),
});
export type RemediateDriftInput = z.infer<typeof RemediateDriftInputSchema>;

export interface RemediateDriftResult {
  diff: StackDiff;
  desiredYaml: string; // serialized desired StackDefinition; empty string if no desired def
  remediable: boolean; // true when status is drift | missing | extra (and desired def exists)
  reason?: string; // why not remediable (e.g., "in_sync" or "no desired state")
}

export const remediateDrift: Tool<RemediateDriftInput, RemediateDriftResult> = {
  name: "remediate_drift",
  description:
    "Detect configuration drift for a stack and return the desired state for remediation. " +
    "The caller (L3) handles confirmation and re-apply.",
  inputSchema: RemediateDriftInputSchema,
  category: "high-level",
  needsPermission: () => true,
  call: async function* (input, ctx): AsyncGenerator<ToolProgress, RemediateDriftResult> {
    yield { type: "progress", msg: `Detecting drift for stack ${input.stackName}...` };
    const diff = await detectDrift(input.stackName, ctx.stateStore, ctx.dockerEngine, ctx.cwd);

    // in_sync: no-op
    if (diff.status === "in_sync") {
      return { diff, desiredYaml: "", remediable: false, reason: "in_sync" };
    }

    // Read the desired definition to serialize
    const def = ctx.stateStore.read(input.stackName);
    if (!def) {
      // no desired definition available (e.g. status="extra" with no recorded definition)
      return { diff, desiredYaml: "", remediable: false, reason: "no desired state" };
    }

    const desiredYaml = stringifyYaml(def);

    // remediable for drift, missing, extra (if def exists)
    return { diff, desiredYaml, remediable: true };
  },
};

File: src/docker_agent/tools/remediate_drift.py
"""remediate_drift tool.

Parity: ``src/tools/remediateDrift.ts``.
"""

from __future__ import annotations

from collections.abc import AsyncIterator

import yaml
from pydantic import BaseModel, Field

from docker_agent.state.drift_detector import detect_drift
from docker_agent.tool import ToolContext, ToolDone, ToolProgress
from docker_agent.types.stack import StackDiff


class RemediateDriftInput(BaseModel):
    stack_name: str = Field(alias="stackName", min_length=1)

    model_config = {"populate_by_name": True}


class RemediateDriftResult(BaseModel):
    diff: StackDiff
    desired_yaml: str = Field(alias="desiredYaml")
    remediable: bool
    reason: str | None = None

    model_config = {"populate_by_name": True}


class _RemediateDriftTool:
    name = "remediate_drift"
    description = (
        "Detect configuration drift for a stack and return the desired state for remediation. "
        "The caller (L3) handles confirmation and re-apply."
    )
    input_schema = RemediateDriftInput
    category = "high-level"

    def needs_permission(self, _input: RemediateDriftInput) -> bool:
        return True

    async def call(
        self, input: RemediateDriftInput, ctx: ToolContext
    ) -> AsyncIterator[ToolProgress | ToolDone]:
        yield ToolProgress(msg=f"Detecting drift for stack {input.stack_name}...")
        diff = await detect_drift(
            input.stack_name, ctx.state_store, ctx.docker_engine, ctx.cwd
        )

        if diff.status == "in_sync":
            yield ToolDone(
                RemediateDriftResult(
                    diff=diff,
                    desired_yaml="",
                    remediable=False,
                    reason="in_sync",
                )
            )
            return

        definition = ctx.state_store.read(input.stack_name)
        if definition is None:
            yield ToolDone(
                RemediateDriftResult(
                    diff=diff,
                    desired_yaml="",
                    remediable=False,
                    reason="no desired state",
                )
            )
            return

        desired_yaml = yaml.safe_dump(
            definition.model_dump(by_alias=True),
            sort_keys=False,
        )
        yield ToolDone(
            RemediateDriftResult(
                diff=diff,
                desired_yaml=desired_yaml,
                remediable=True,
            )
        )


remediate_drift = _RemediateDriftTool()

__all__ = ["RemediateDriftInput", "RemediateDriftResult", "remediate_drift"]